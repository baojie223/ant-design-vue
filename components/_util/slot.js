import Vue from 'vue';

// 获取插槽
export function getSlot(vm, name = 'default', props = {}) {
  if (!(vm instanceof Vue)) throw Error('Please input vue instance');
  const nodes1 = (vm.$scopedSlots[name] && vm.$scopedSlots[name](props)) || [];
  const nodes2 = vm.$slots[name] || [];
  if (nodes1.length && nodes2.length) {
    return [...nodes1, ...nodes2];
  } else if (nodes1.length) {
    return [...nodes1];
  } else if (nodes2.length) {
    return [...nodes2];
  } else {
    return [];
  }
}

// 判断是否有插槽
export function hasSlot(vm, name = 'default', props = {}) {
  if (!(vm instanceof Vue)) throw Error('Please input vue instance');
  const nodes1 = (vm.$scopedSlots[name] && vm.$scopedSlots[name](props)) || [];
  const nodes2 = vm.$slots[name] || [];
  if (nodes1.length || nodes2.length) {
    return true;
  } else {
    return false;
  }
}
