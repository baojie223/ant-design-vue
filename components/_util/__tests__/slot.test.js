import Vue from 'vue';
import { mount } from '@vue/test-utils';
import { asyncExpect } from '@/tests/utils';
import { getSlot, hasSlot } from '../slot';

const Father = new Vue({
  render() {
    return <Child scopedSlots={{ default: () => <h1>h1: default</h1> }}></Child>;
  },
});

const Child = new Vue({
  render() {
    return (
      <div>
        {getSlot(this)}
        <span>{hasSlot(this)}</span>
      </div>
    );
  },
});

describe('getSlot', () => {
  it('should has default slot', async () => {
    const wrapper = mount(
      {
        render() {
          return <Father></Father>;
        },
      },
      { sync: false },
    );
    await asyncExpect(() => {
      expect(wrapper.html()).toMatchSnapshot();
    });
  });
});
