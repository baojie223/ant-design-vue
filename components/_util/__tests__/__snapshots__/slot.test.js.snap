// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getSlot should has default slot 1`] = `undefined`;
